src/sidebar/components/markdown-editor.js

```diff
@@ -0,0 +1,291 @@
+    if (
+      before.length === 0 ||
+      before.slice(-1) === '\n' ||
+      before.slice(-2) === '$$'
+    ) {
+      return toggleSpanStyle(state, '$$', '$$', 'Insert LaTeX');
+    } else {
+      return toggleSpanStyle(state, '\\(', '\\)', 'Insert LaTeX');
+    }
+  };
+
+  switch (command) {
+    case 'bold':
+      update(state => toggleSpanStyle(state, '**', '**', 'Bold'));
+      break;
+    case 'italic':
+      update(state => toggleSpanStyle(state, '*', '*', 'Italic'));
+      break;
+    case 'quote':
+      update(state => toggleBlockStyle(state, '> '));
+      break;
+    case 'link':
+      update(state => convertSelectionToLink(state));
+      break;
+    case 'image':
+      update(state => convertSelectionToLink(state, LinkType.IMAGE_LINK));
+      break;
+    case 'math':
+      update(insertMath);
+      break;
+    case 'numlist':
+      update(state => toggleBlockStyle(state, '1. '));
+      break;
+    case 'list':
+      update(state => toggleBlockStyle(state, '* '));
+      break;
+    default:
+      throw new Error(`Unknown toolbar command "${command}"`);
+  }
+}
+
+/**
+ * Viewer/editor for the body of an annotation in markdown format.
+ */
+function MarkdownEditor({ onEditText = () => {}, text = '' }) {
+  /** Whether the preview mode is currently active. */
+  const [preview, setPreview] = useState(false);
+
+  /** The input element where the user inputs their comment. */
+  const input = useRef(null);
+
+  useEffect(() => {
+    if (!preview) {
+      input.current.focus();
+    }
+  }, [preview]);
+
+  const togglePreview = () => setPreview(!preview);
+  const handleCommand = command => {
+    const inputEl = input.current;
+    handleToolbarCommand(command, inputEl);
+    onEditText({ text: inputEl.value });
+  };
+
+  const handleKeyDown = event => {
+    if (!event.ctrlKey) {
+      return;
+    }
+
+    for (let [command, key] of Object.entries(SHORTCUT_KEYS)) {
+      if (key === event.key) {
+        event.stopPropagation();
+        event.preventDefault();
+        handleCommand(command);
+      }
+    }
+  };
+
+  return (
+    <div>
+      <Toolbar
+        onCommand={handleCommand}
+        isPreviewing={preview}
+        onTogglePreview={togglePreview}
+      />
+      {preview ? (
+        <MarkdownView
+          textClass={{ 'markdown-editor__preview': true }}
+          markdown={text}
+        />
+      ) : (
+        <textarea
+          className="form-input form-textarea"
+          ref={input}
+          onClick={e => e.stopPropagation()}
+          onKeydown={handleKeyDown}
+          onInput={e => onEditText({ text: e.target.value })}
+          value={text}
+        />
+      )}
+    </div>
+  );
+}
+
+MarkdownEditor.propTypes = {
+  /** The markdown text to edit. */
+  text: propTypes.string,
+
+  /**
+   * Callback invoked with `{ text }` object when user edits text.
+   *
+   * TODO: Simplify this callback to take just a string rather than an object
```

Comment:
How would a `string` callback work? :)

Reply:
The signature of the callback is currently `(value: { text: string }) => any` (ie. the argument is an object with a `text` field. I'm suggesting it should just be `(text: string) => any` (ie. the argument should just be a string).